### ISSUE TITLE
Explore SourceKitten as an opportunity for parsing Objective-C code

### ISSUE THREAD

**Issue by DanielMSchmidt** [wix/Detox#688]
We can translate the output of SourceKitten over to the structure we need:

- Classes can be found as "key.kind" : "sourcekitten.source.lang.objc.decl.protocol",
- Methods can be found as "key.kind" : "sourcekitten.source.lang.objc.decl.method.instance",
	- Methodname is "key.name"
	- Comment is "key.doc.comment"
	- Parameters are key.doc.parameters
		- Name of a parameter: name
		- Type of a parameter: ???
	- Return type: ???

**Comment by LeoNatan** [comment]
I think return type and maybe param types are empty when the type is `id`.

**Comment by DanielMSchmidt** [comment]
Could be, will try what happens with non-id types. Although we normally need the id typed information, too, so that we can validate that we really get an action or a matcher as parameter

**Comment by LeoNatan** [comment]
`id` is a "anyobject" type, so it really doesn't matter. By default in Objective C, if type is not provided, it is assumed to be `id`.

So for

```objc
@interface ZZ : NSObject

- someSelector:x;

@end
```

Both here are `id`.

**Comment by DanielMSchmidt** [comment]
I opened an issue on their side to get a bit more information on this: https://github.com/jpsim/SourceKitten/issues/516

**Comment by rotemmiz** [comment]
@DanielMSchmidt, we still can't upgrade EarlGrey version (generation scripts fail to generate code for 1.15.0). 
Can we switch objective-c-parser to user source-kitten ? We need this as fast as possible.

**Comment by rotemmiz** [comment]
For now we can fix the generation failure with this PR https://github.com/DanielMSchmidt/objective-c-parser/pull/89

**Comment by DanielMSchmidt** [comment]
Sourcekitten would require a lot of work within sourcekitten itself, I am therefore exploring other tools, like [ANRLR](http://www.antlr.org/), which also has a lot of [grammars](https://github.com/antlr/grammars-v4) already premade. I will do a spike to verify that this will be a faster path.

**Comment by LeoNatan** [comment]
Depending on how robust it is, it could be a good alternative. Thanks!

**Comment by LeoNatan** [comment]
Does it know to recursive follow header imports/includes?

**Comment by DanielMSchmidt** [comment]
It can't follow, that's the downside. Is this really a hard retirement for us?

**Comment by LeoNatan** [comment]
I think so. Stuff moves around all the time, but not necessarily renamed. We need to follow the headers.

**Comment by DanielMSchmidt** [comment]
I did some investigation today and I think I can totally ship a new version without the stuff moving considered. The good thing would be I could ideally use the same implementation for both iOS and Android just exchanging the underlying grammar. So this would be my roadmap:

1. Deliver an exchange for the objective-c module to reduce the risk of updates killing the generation process
2. Move the android side over to this parser type based on grammars to be sure we are on the save side here
3. Find a solid way to follow-up on references. Finding them isn't the problem, but the resolving side might be. I am not sold on any solution here, so I will do some research before starting with this work.

@LeoNatan Do you think this is a good path moving forward? I would love to hear your input on this 👍

**Comment by DanielMSchmidt** [comment]
Thank you for your input @LeoNatan, I haven't thought about some of these aspects.
Thinking about it I see a split decision between two different types of benefits, so let me try to summarize the two options we have now:

### Grammar-based

**Pro**

- Can handle both Objective-C and Java
- Is much faster & easier to develop

**Cons**

- We will need to do manual work within detox (defining aliases for type aliases, moving code around whenever the structure of their source code changes)
- If the grammar fails for some reason we need to make adjustments to it (not very likely, it seems battle tested)

### Compilation-AST-based

**Pro**

- We see the same thing as the source code, therefore there can't be any manual work required

**Cons**

- We need to write two solutions
- From my current point of view it seems a lot harder (it would at least require one or two upstream PRs to source kitten and a lot of deep objective-c environment knowledge)


I understood it like we need to have a better solution ASAP, this is why I would go with the grammar based approach. I agree that the "perfect" and not at all manual solution would be the compilation AST based solution, so if we don't need something better soon, but rather something great in the long run I would go with this solution.

**Comment by LeoNatan** [comment]
Two solutions is a feature, not a bug 😉

I’m going on vacation. When I return, I may take a look at SourceKitten to see what is missing.

**Comment by LeoNatan** [comment]
Two and a half weeks.

**Comment by DanielMSchmidt** [comment]
I could have a grammar-based solution until then, but on the other hand that might be wasted time.

**Comment by LeoNatan** [comment]
This has not advanced as we want it to, and it seems like out of scope at this point. Closing.
